src/pages/instances/StoragePoolMigrationTable.tsx

```diff
@@ -0,0 +1,244 @@
+import { FormDevice } from "util/formDevices";
+
+interface Props {
+  close: () => void;
+  instance: LxdInstance;
+  onSelect: (pool: string) => void;
+  targetPool: string;
+}
+
+const StoragePoolMigrationTable: FC<Props> = ({
+  close,
+  instance,
+  onSelect,
+  targetPool,
+}) => {
+  const toastNotify = useToastNotification();
+  const instanceLoading = useInstanceLoading();
+  const eventQueue = useEventQueue();
+  const queryClient = useQueryClient();
+  const { data: pools = [] } = useQuery({
+    queryKey: [queryKeys.storage],
+    queryFn: () => fetchStoragePools(instance.project),
+  });
+
+  const handleSuccess = (newTarget: string, instanceName: string) => {
+    toastNotify.success(
+      <>
+        Instance <ItemName item={{ name: instanceName }} bold /> successfully
+        migrated to storage pool <ItemName item={{ name: newTarget }} bold />
+      </>,
+    );
+  };
+
+  const notifyFailure = (e: unknown, instanceName: string) => {
+    instanceLoading.setFinish(instance);
+    toastNotify.failure(
+      `Storage pool migration failed on instance ${instanceName}`,
+      e,
+    );
+  };
+
+  const handleFailure = (msg: string, instanceName: string) => {
+    notifyFailure(new Error(msg), instanceName);
+  };
+
+  const handleFinish = () => {
+    void queryClient.invalidateQueries({
+      queryKey: [queryKeys.instances, instance.name],
+    });
+    instanceLoading.setFinish(instance);
+  };
+
+  const handleMigrate = () => {
+    instanceLoading.setLoading(instance, "Migrating");
+    migrateInstance(instance.name, instance.project, undefined, targetPool)
+      .then((operation) => {
+        eventQueue.set(
+          operation.metadata.id,
+          () => handleSuccess(targetPool, instance.name),
+          (err) => handleFailure(err, instance.name),
+          handleFinish,
+        );
+        toastNotify.info(
+          `Storage pool migration started for instance ${instance.name}`,
+        );
+        void queryClient.invalidateQueries({
+          queryKey: [queryKeys.instances, instance.name, instance.project],
+        });
+      })
+      .catch((e) => {
+        notifyFailure(e, instance.name);
+      })
+      .finally(() => {
+        close();
+      });
+  };
+
+  const headers = [
+    { content: "Name", sortKey: "name" },
+    { content: "Driver", sortKey: "driver" },
+    { content: "Size", className: "size" },
+    { content: "Status", sortKey: "status" },
+    {
+      content: "Volumes",
+      sortKey: "volumes",
+      className: "u-align--right",
+    },
+    { "aria-label": "Actions", className: "actions" },
+  ];
+
+  const rows = pools.map((pool) => {
+    const volumes =
+      pool.used_by?.filter((url) => !url.startsWith("/1.0/profiles")) ?? [];
+    const totalVolumeCount = volumes.length;
+    const rootDiskDevice = Object.values(instance.expanded_devices ?? {}).find(
+      (device) => isRootDisk(device as FormDevice),
+    ) as LxdDiskDevice;
```

Comment:
I think we should move this out of the loop.

Reply:
good catch thanks!